r/R/dotplot.R

```diff
@@ -1,35 +1,88 @@
+  group_by_cell_sets <- req$body$groupBy$children
+  filter_by <- req$body$filterBy
+  is_filtered_by_all <- req$body$isFilterByAll
 
-  for (i in seq_along(cell_sets)) {
-    set <- cell_sets[[i]]
-    filtered_cells <- intersect(set$cellIds, cells_id)
-    data$custom[cells_id %in% filtered_cells] <- i
+  if (length(group_by_cell_sets) < 1) {
+    message("The requested Cell Sets are empty. Returning empty results.")
+    return(list())
   }
 
-  if (type == "custom") {
-    annot <- data@misc$gene_annotations
-    annot_subset <- subset(annot, toupper(name) %in% toupper(req_genes))
-    annot_subset <- annot_subset[, c("input", "name")]
+  #Collect ids to subset object
+  if (!is_filtered_by_all) {
+    subsetIds <- filter_by$cellIds
   } else {
-    # type == marker
-    # Need the marker features
+    subsetIds <- list()
+    for (i in seq_along(filter_by)) {
+      filter_parent <- filter_by[[i]]$children
+      for(j in seq_along(filter_parent)) {
+        subsetIds <- append(subsetIds, filter_parent[[j]]$cellIds)
+      }
+    }
   }
 
-  dotplot_data <- Seurat::DotPlot(data, features = annot_subset$input, group.by = "custom")$data
-  #features.plot has the ensemble ids
-  dotplot_data$name <- annot_subset[dotplot_data$features.plot, "name"]
+  #Subset cell Ids
+  if (length(subsetIds)) {
+    meta_data_subset <- data@meta.data[match(subsetIds, data@meta.data$cells_id), ]
+    current_cells <- rownames(meta_data_subset)
+    data <- subset(data, cells = current_cells)
+    cells_id <- data$cells_id
+  } else {
+    message("The ids to subset the object are empty. Returning empty results.")
+    return(list())
+  }
 
-  dotplot_data <- dotplot_data[, c("avg.exp", "pct.exp", "name", "id")]
-  colnames(dotplot_data) <- c("avgExpression", "cellsPercentage", "geneName", "cellSet")
+  #Construct the custom slot
+  for (i in seq_along(group_by_cell_sets)) {
+    cell_set <- group_by_cell_sets[[i]]
+    filtered_cells <- intersect(cell_set$cellIds, cells_id)
+    if (cell_set$name %in% data$custom) {
+      data$custom[cells_id %in% filtered_cells] <- paste0(cell_set$name, i)
+    } else {
+      data$custom[cells_id %in% filtered_cells] <- cell_set$name
+    }
+  }
+  # If NA values are left in the group, dotplot function will fail.
+  data <- subset(data, subset = custom != "NA")
 
-  res <- purrr::transpose(dotplot_data)
+  #Get marker genes or requested gene names.
+  if (useMarkerGenes) {
+    num_features <- req$body$numberOfMarkers
+    all_markers <- getTopMarkerGenes(num_features, data, group_by_cell_sets)
+    features <- as.data.frame(getMarkerNames(data, all_markers))
+    rownames(features) <- features$input
+  } else {
+    req_genes <- req$body$customGenesList
+    annot <- data@misc$gene_annotations
+    annot_subset <- subset(annot, toupper(name) %in% toupper(req_genes))
+    features <- annot_subset[, c("input", "name")]
+  }
+
+  dotplot_data <- Seurat::DotPlot(data, features = features$input, group.by = "custom")$data
```

Comment:
why do you have to do `features = features$input` in here, when in line 72 (`rownames(features) <- features$input`) you've already kind of done that?